src/tree/updater_quantile_hist.cc

```diff
@@ -88,94 +89,233 @@ bool QuantileHistMaker::UpdatePredictionCache(
   }
 }
 
-void QuantileHistMaker::Builder::SyncHistograms(
-    int starting_index,
-    int sync_count,
-    RegTree *p_tree) {
-  builder_monitor_.Start("SyncHistograms");
-  this->histred_.Allreduce(hist_[starting_index].data(), hist_builder_.GetNumBins() * sync_count);
-  // use Subtraction Trick
-  for (auto const& node_pair : nodes_for_subtraction_trick_) {
-    hist_.AddHistRow(node_pair.first);
-    SubtractionTrick(hist_[node_pair.first], hist_[node_pair.second],
-                     hist_[(*p_tree)[node_pair.first].Parent()]);
-  }
-  builder_monitor_.Stop("SyncHistograms");
+void QuantileHistMaker::Builder::BuildNodeStat(
+    const GHistIndexMatrix &gmat,
+    DMatrix *p_fmat,
+    RegTree *p_tree,
+    const std::vector<GradientPair> &gpair_h,
+    int32_t nid) {
+
+  // add constraints
+  if (!(*p_tree)[nid].IsLeftChild() && !(*p_tree)[nid].IsRoot()) {
+    auto parent_id = (*p_tree)[nid].Parent();
+    // it's a right child
+    auto left_sibling_id = (*p_tree)[parent_id].LeftChild();
+    auto parent_split_feature_id = snode_[parent_id].best.SplitIndex();
+
+    {
```

Comment:
I don't think we need an extra layer of scoping here. Can you put `spliteval_->AddSplit(*)` in the same level as the rest of code?

Reply:
Done